### PR TITLE
Deleted duplicate swagger settings and fixed users tests

### PR DESCRIPTION
**Pull request by danylkoltak**

### CODE REVIEW

**Review by SergeKry — APPROVED** [review]

**Review by devdsys — APPROVED** [review]

File: forum/users/tests.py
Comment (mehalyna) [inline forum/users/tests.py:1]: Using cleanup in the `tearDown` method can be better approach to ensure that each test starts with a clean state without affecting the performance of the `setUp` method. 
```
from django.test import TestCase
from django.urls import reverse
from users.models import CustomUser
from users.tests.factories import CustomUserFactory

class UserRegistrationTests(APITestCase):
    def setUp(self):
        self.register_url = reverse('auth:user-register')
        self.valid_payload = self.create_payload()
        self.user = CustomUserFactory.create()  # Use a factory to create a user

    def tearDown(self):
        CustomUser.objects.all().delete()  # Cleanup after each test

    def create_payload(self):
        return {
            "email": "testuser@example.com",
            "password": "securepassword",
            "role": "INVESTOR"
        }
```

Comment (mehalyna) [inline forum/users/tests.py:1]: In case if you you have convictions гsing cleanup in the `setUp` method, there are some considerations and potential improvements to this approach:

1. Django's `TestCase` (which `APITestCase` inherits from) already wraps each test in a transaction and rolls it back at the end, ensuring a clean state. Rely on this feature instead of manually deleting records.
2. Use factory methods or fixtures to create the necessary data for each test. This ensures that each test has the data it needs without relying on the state of the database.
3. If you must delete records, consider doing it selectively, targeting only the records created during the test.

_Example:_
```
from django.test import TestCase
from django.urls import reverse
from users.models import CustomUser
from users.tests.factories import CustomUserFactory

class UserRegistrationTests(APITestCase):
    def setUp(self):
        self.register_url = reverse('auth:user-register')
        self.valid_payload = self.create_payload()
        self.user = CustomUserFactory.create()  # Use a factory to create a user

    def create_payload(self):
        return {
            "email": "testuser@example.com",
            "password": "securepassword",
            "role": "INVESTOR"
        }
```

**Review by mehalyna — CHANGES_REQUESTED** [review]

File: forum/users/serializers.py [hunk 1]
```diff
@@ -82,15 +84,37 @@ def validate_role(self, value):
...
 
-    def validate_last_name(self, value):
-        if len(value) > 30:
-            raise serializers.ValidationError("Last name must not exceed 30 characters.")
-        return value
+VALID_TOKEN_ROLES = [Role.STARTUP, Role.INVESTOR]
+
+
+class CustomTokenObtainPairSerializer(TokenObtainPairSerializer):
+    role = serializers.ChoiceField(choices=[(role.value, role.name.lower()) for role in VALID_TOKEN_ROLES])
```
Comment (mehalyna) [inline forum/users/serializers.py:92]: The role validation logic is written in the `validate` method. Moving this to the `ChoiceField` directly improves readability and reusability.
```
role = serializers.ChoiceField(choices=[(role.value, role.name.lower()) for role in VALID_TOKEN_ROLES], required=True)
```

File: forum/users/serializers.py [hunk 2]
```diff
@@ -82,15 +84,37 @@ def validate_role(self, value):
...
+
+        if role is None:
+            raise exceptions.ValidationError({"role": "This field is required."})
+        
+        try:
+            role = int(role)
+            if role not in [role.value for role in VALID_TOKEN_ROLES]:
+                raise ValueError
+        except (ValueError, TypeError):
+            raise exceptions.ValidationError({"role": "Invalid role specified."})
```
Comment (mehalyna) [inline forum/users/serializers.py:108]: `ValueError` and `TypeError` are caught generically. This could obscure debugging.
Provide specific error messages and log the invalid inputs.
```
try:
    role = int(role)
    if role not in [role.value for role in VALID_TOKEN_ROLES]:
        raise ValueError(f"Invalid role: {role}")
except ValueError as e:
    raise exceptions.ValidationError({"role": str(e)})
except TypeError:
    raise exceptions.ValidationError({"role": "Role must be an integer."})
```

File: forum/users/tests/test_permissions.py
Comment (mehalyna) [inline forum/users/tests/test_permissions.py:1]: 1. The test cases are lengthy and could be simplified by breaking down helper functions and making assertions more explicit. Use helper methods and better organize test descriptions.
2. Add tests to handle edge cases, such as:
 - Role mismatches (e.g., a role that is not part of `VALID_TOKEN_ROLES`).
 - Null or empty role in the token.
3. The setup uses repetitive code for mocking user roles and tokens. Consolidate setup logic into reusable helpers.

File: forum/users/tests/test_token_serializer.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+
+class TestCustomTokenObtainPairSerializer(APITestCase):
+    def setUp(self):
+        # Create a test user
+        self.user = User.objects.create_user(
+            email="testuser@example.com",
+            password="securepassword",
+            role=Role.INVESTOR.value  # Assign a default role
+        )
```
Comment (mehalyna) [inline forum/users/tests/test_token_serializer.py:19]: Consider using `setUpTestData` instead of `setUp` if the test data does not change between tests. This can improve performance by creating the test data once for the entire TestCase.
 ```
@classmethod
def setUpTestData(cls):
    cls.user = User.objects.create_user(
        email=TEST_EMAIL,
        password=TEST_PASSWORD,
        role=Role.INVESTOR.value
    )
```

File: forum/users/tests/test_token_serializer.py
Comment (mehalyna) [inline forum/users/tests/test_token_serializer.py:1]: The `_get_mock_request` method is useful, but you might want to try Django's `RequestFactory` for creating mock requests, which can provide more realistic request objects.
```
from django.test import RequestFactory

def _get_mock_request(self, data):
    factory = RequestFactory()
    request = factory.post('/fake-url/', data, content_type='application/json')
    return request
```

File: forum/users/tests/test_token_serializer.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        self.assertEqual(refresh_token["role"], Role.STARTUP.value)
+
+    def test_invalid_role(self):
+        """Test that an invalid role raises a ValidationError."""
+        request_data = {"email": self.user.email, "password": "securepassword", "role": 9999}  # Invalid role
+        context = {"request": self._get_mock_request(request_data)}
+
+        serializer = CustomTokenObtainPairSerializer(context=context)
+        with self.assertRaises(ValidationError) as exc:
+            serializer.validate(request_data)
```
Comment (mehalyna) [inline forum/users/tests/test_token_serializer.py:41]: When checking for exceptions, you can use `assertRaisesMessage` to ensure the exception message is as expected.
```
with self.assertRaisesMessage(ValidationError, "Invalid role specified."):
    serializer.validate(request_data)
```

File: forum/users/tests/test_token_serializer.py
Comment (mehalyna) [inline forum/users/tests/test_token_serializer.py:1]: Ensure you have tests for edge cases, such as:
 - Invalid email or password.
 - Role values that are valid but not assigned to any user.

**Review by mehalyna — CHANGES_REQUESTED** [review]
